fdbserver/IPager.h

```diff
@@ -97,67 +151,341 @@ class ArenaPage : public ReferenceCounted<ArenaPage>, public FastAllocated<Arena
-			ASSERT(p->size() == usableSize);
-			memcpy(wptr, p->begin(), usableSize);
-			wptr += usableSize;
+	// The next two functions set mostly forensic info that may help in an investigation to identify data on disk.  The
+	// exception is pageID which must be set to the physical page ID on disk where the page is written or post-read
+	// verification will fail.
+	void setWriteInfo(PhysicalPageID pageID, Version writeVersion) {
+		if (page->headerVersion == 1) {
+			HeaderV1* h = page->getMainHeader<HeaderV1>();
+			h->firstPhysicalPageID = pageID;
+			h->writeVersion = writeVersion;
+			h->writeTime = now();
+		}
+	}
+	void setLogicalPageInfo(LogicalPageID lastKnownLogicalPageID, LogicalPageID lastKnownParentLogicalPageID) {
+		if (page->headerVersion == 1) {
+			HeaderV1* h = page->getMainHeader<HeaderV1>();
+			h->lastKnownLogicalPageID = lastKnownLogicalPageID;
+			h->lastKnownParentLogicalPageID = lastKnownParentLogicalPageID;
 		}
-
-		return Reference<ArenaPage>(superpage);
 	}
 
-	Checksum& getChecksum() { return *(Checksum*)(buffer + size()); }
+	// Must be called before writing to disk to update headers and encrypt page
+	// Pre:   Encoding-specific header fields are set if needed
+	//        Secret is set if needed
+	// Post:  Main and Encoding subheaders are updated
+	//        Payload is possibly encrypted
+	void preWrite(PhysicalPageID pageID) const {
+#if VALGRIND
+		// Explicitly check payload definedness to make the source of valgrind errors more clear.
+		// Without this check, calculating a checksum on a payload with undefined bytes does not
+		// cause a valgrind error but the resulting checksum is undefined which causes errors later.
+		ASSERT(VALGRIND_CHECK_MEM_IS_DEFINED(pPayload, payloadSize) == 0);
+#endif
+		if (page->encodingType == EncodingType::XXHash64) {
+			page->getEncodingHeader<XXHashEncodingHeader>()->encode(pPayload, payloadSize, pageID);
+		} else if (page->encodingType == EncodingType::XOREncryption) {
+			ASSERT(secret.size() == 1);
+			page->getEncodingHeader<XOREncryptionEncodingHeader>()->encode(secret[0], pPayload, payloadSize, pageID);
+		} else {
+			throw page_encoding_not_supported();
+		}
 
-	Checksum calculateChecksum(LogicalPageID pageID) { return XXH3_64bits_withSeed(buffer, size(), pageID); }
+		if (page->headerVersion == 1) {
+			page->getMainHeader<HeaderV1>()->update(buffer, pPayload - buffer);
+		} else {
+			throw page_header_version_not_supported();
+		}
+	}
 
-	void updateChecksum(LogicalPageID pageID) { getChecksum() = calculateChecksum(pageID); }
+	// Must be called after reading from disk to verify all non-payload bytes
+	// Pre:   Bytes from storage medium copied into raw buffer space
+	// Post:  Page headers outside of payload are verified (unless verify is false)
+	//        Payload is accessible via data(), dataSize(), etc.
+	//
+	// Exceptions are thrown for unknown header types or pages which fail verification
+	void postRead1(PhysicalPageID pageID, bool verify = true) {
```

Comment:
Maybe postReadHeader and postReadPayload would be more descriptive function names than postRead1/2?

Reply:
+1

Reply:
Good idea, will do.